### PR TITLE
#332 Use UTBotCpp as static analyzer and report results in SARIF

### PR DESCRIPTION
**Pull request by alexey-utkin**
- SARIF generation in server and CLI modes
- auto open SARIF result in MS plugin after the run, if MS plugin has been installed
- backup old SARIF reports with creation-date timestamp in name

### CODE REVIEW

**Review by sava-cska — CHANGES_REQUESTED** [review]

File: server/test/suites/cli/goldenImage/codeAnalysis/project_code_analysis.sarif [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+      "tool": {
+        "driver": {
+          "informationUri": "https://utbot.org",
+          "name": "UTBotCpp"
+        }
+      }
+    }
+  ],
+  "version": "2.1.0"
+}
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/test/suites/cli/goldenImage/codeAnalysis/project_code_analysis.sarif:242]: It is auto-generated. So as-is.

File: server/src/streams/tests/ServerTestsWriter.cpp [hunk 2]
```diff
@@ -60,3 +64,24 @@ bool ServerTestsWriter::writeFileAndSendResponse(const tests::Tests &tests,
...
+        testSource->set_filepath(pathToStore);
+        if (synchronizeCode) {
+            testSource->set_code(content);
+        }
+        LOG_S(INFO) << message;
+        auto progress = GrpcUtils::createProgress(message, 100, false);
+        response.set_allocated_progress(progress.release());
+        writeMessage(response);
+    }
+}
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/streams/tests/ServerTestsWriter.cpp]: fixed

File: server/src/printers/TestsPrinter.cpp [hunk 3]
```diff
@@ -112,18 +171,31 @@ void TestsPrinter::genCode(Tests::MethodDescription &methodDescription,
...
+    return printer::Printer::concat(testBaseName, "_test_", testNum);
+}
+
 void TestsPrinter::genCodeBySuiteName(const std::string &targetSuiteName,
                                       Tests::MethodDescription &methodDescription,
                                       const std::optional<LineInfo::PredicateInfo>& predicateInfo,
                                       bool verbose,
                                       int &testNum) {
-    const auto& testCases = methodDescription.suiteTestCases[targetSuiteName];
+    auto& testCases = methodDescription.suiteTestCases[targetSuiteName];
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/printers/TestsPrinter.cpp]: fixed

**Review by sava-cska — CHANGES_REQUESTED** [review]

File: server/src/KleeRunner.cpp [hunk 4]
```diff
@@ -44,7 +45,9 @@ void KleeRunner::runKlee(const std::vector<tests::TestMethod> &testMethods,
         fileToMethods[method.sourceFilePath].push_back(method);
     }
 
-    std::function<void(tests::Tests &tests)> writeFunctor = [&](tests::Tests &tests) {
+    nlohmann::json sarifResults;
+
+    std::function<void(tests::Tests &tests)> prepateTests = [&](tests::Tests &tests) {
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/KleeRunner.cpp]: fixed

**Review by ladisgin — CHANGES_REQUESTED** [review]

File: server/src/KleeRunner.cpp [hunk 5]
```diff
@@ -146,7 +166,22 @@ static void processMethod(MethodKtests &ktestChunk,
                             return UTBotKTestObject{ kTestObject };
                         });
 
-                    ktestChunk[method].emplace_back(objects, status);
+                    std::vector<std::string> errorDescriptors = CollectionUtils::transform(
+                        errorDescriptorFiles, [](const fs::path &errorFile) {
+                            std::ifstream fileWithError(errorFile.c_str(), std::ios_base::in);
+                            std::string content((std::istreambuf_iterator<char>(fileWithError)),
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/KleeRunner.cpp:172]: No. Please read https://en.wikipedia.org/wiki/Most_vexing_parse
https://stackoverflow.com/questions/2912520/read-file-contents-into-a-string-in-c

File: server/src/printers/TestsPrinter.cpp [hunk 6]
```diff
@@ -59,7 +61,64 @@ void TestsPrinter::joinToFinalCode(Tests &tests, const fs::path& generatedHeader
...
+                int testIndex = -1;
+                try {
+                    testIndex = std::stoi(nameAndTestIndex.substr(pos + 1));
+                } catch (std::logic_error &e) {
+                    // ignore
+                }
+                Tests::MethodsMap::iterator it = tests.methods.find(name);
+                if (it != tests.methods.end() && testIndex >= 0) {
+                    Tests::MethodDescription &methodDescription = it.value();
+                    std::vector<Tests::MethodTestCase> &testCases =methodDescription.testCases;
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/printers/TestsPrinter.cpp:92]: fixed

File: server/src/printers/TestsPrinter.cpp [hunk 7]
```diff
@@ -59,7 +61,64 @@ void TestsPrinter::joinToFinalCode(Tests &tests, const fs::path& generatedHeader
...
+                        Tests::MethodTestCase &testCase = testCases[testIndex];
+                        auto &descriptors = testCase.errorDescriptors;
+                        if (testCase.errorDescriptors.empty()) {
+                            LOG_S(ERROR) << "no error info for test case: "
+                                         << name
+                                         << ", test #"
+                                         << testIndex;
+                            continue;
+                        }
+                        std::stringstream ss;
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/printers/TestsPrinter.cpp:103]: fixed

File: server/src/SARIFGenerator.cpp [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+                                const std::regex stack_regex(
+                                    R"regex(\s+#(.*) in ([^ ]*) [(][^)]*[)] at ([^:]*):(\d+))regex");
+                                std::smatch stack_match;
+                                if (!std::regex_match(lineInDescriptor, stack_match, stack_regex)) {
+                                    LOG_S(ERROR) << "wrong `Stack` line: " << lineInDescriptor;
+                                } else {
+                                    const fs::path &srcPath = fs::path(stack_match[3]);
+                                    const fs::path &relPathInProject = getInProjectPath(projectContext.projectPath, srcPath);
+                                    const fs::path &fullPathInProject = projectContext.projectPath / relPathInProject;
+                                    if (Paths::isSubPathOf(projectContext.buildDir, fullPathInProject)) {
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/SARIFGenerator.cpp:76]: Seems we would like to skip any non-user code. Let it be as it is.

File: server/src/SARIFGenerator.cpp [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+namespace sarif {
+    // Some more logic above standard implementation
+    //     `fs::relative(src, path)`
+    // for the case of src as the relative path.
+    // For example:
+    //     getInProjectPath("/home/utbot/tmp/UTBotCpp/server/test/suites/object-file",
+    //                      "test/suites/object-file/op/source2.c")
+    // returns
+    //     "op/source2.c"
+    fs::path getInProjectPath(const fs::path &path, const fs::path &src) {
```
[the remark this answers is not in the document]
Comment (alexey-utkin) [reply server/src/SARIFGenerator.cpp:24]: fixed

**Review by ladisgin — APPROVED** [review]
squash commits

**Review by sava-cska — APPROVED** [review]

**Pull request merged**